### PR TITLE
[aes,SiVal] Add SiVal test plans

### PR DESCRIPTION
**Pull request by vrozic**
This PR adds a list of features to `aes.hjson`, moves AES testplan into a separate file and extends existing tests to cover more features.

### CODE REVIEW

**Review by marnovandermaas — COMMENTED** [review]
From first glance this looks fine to me. Would you mind splitting the test plan move into a separate commit from where you add the features and sival stages?

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vogelpi) [inline hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:21]: Can you explain what the motivation is to run this test in all these LC states? Do we really need this?

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vogelpi) [inline hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:74]: I don't think we need to run this in all LC states but maybe I misunderstood something here. Can you please explain?

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vogelpi) [inline hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:73]: SV2 -> SV3? Or SV1 because the test is very basic?

**Review by vogelpi — CHANGES_REQUESTED** [review]
Thanks @vrozic for working on this.

I've left some comments. In general the following should be reworked:
1. All tests seem to have gotten the same SiVal stage (SV2) and all / almost all LC states. Is this intentional? I don't think this makes sense.
2. Some of the features are not tested / not linked to the tests. To me it seems that in some cases, features got forgotten. In other cases, we need to extend the tests.
3. Some aspects of AeS we discussed previously have not made it into the SiVal feature list/testplan. Can you please check the notes and extend accordingly?

File: hw/ip/aes/data/aes.hjson
Comment (vrozic) [reply hw/ip/aes/data/aes.hjson:242]: I can split AES.MODE.ECB into three features.
Should we also split the other modes by key length, or would that be too many features?

Comment (vrozic) [reply hw/ip/aes/data/aes.hjson:242]: Agreed, I'll change all to AES

File: hw/ip/aes/data/aes.hjson
Comment (vrozic) [reply hw/ip/aes/data/aes.hjson:251]: Mistake when copying. I'm fixing it now.

File: hw/ip/aes/data/aes.hjson
Comment (vrozic) [reply hw/ip/aes/data/aes.hjson:295]: I've probably misunderstood the documentation about this one. I'll correct it

File: hw/ip/aes/data/aes.hjson
Comment (vrozic) [reply hw/ip/aes/data/aes.hjson:291]: My mistake. I thought that `chip_sw_aes_masking_off` can test this feature.
For me it's OK to remove this if it's not testable

File: hw/ip/aes/data/aes.hjson
Comment (vogelpi) [reply hw/ip/aes/data/aes.hjson:242]: Ah no, you don't need to split the modes. Key length and mode are completely orthogonal.

Comment (vrozic) [reply hw/ip/aes/data/aes.hjson]: I see.
I've now added key lengths as separate features, so we have 3 KEY_LEN features and 5 MODE features.
I assume that the test plan should include all 15 combinations.

File: hw/ip/aes/data/aes.hjson
Comment (vrozic) [reply hw/ip/aes/data/aes.hjson]: Fixed

File: hw/ip/aes/data/aes.hjson
Comment (vrozic) [reply hw/ip/aes/data/aes.hjson]: Changed it to 'The key...'

File: hw/ip/aes/data/aes.hjson
Comment (vrozic) [reply hw/ip/aes/data/aes.hjson]: Fixed

File: hw/ip/aes/data/aes.hjson
Comment (vrozic) [reply hw/ip/aes/data/aes.hjson]: fixed

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vrozic) [reply hw/top_earlgrey/data/ip/chip_aes_testplan.hjson]: I see. I've changed it to SV1 now

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vrozic) [reply hw/top_earlgrey/data/ip/chip_aes_testplan.hjson]: I see. I've removed it now.

File: hw/ip/aes/data/aes.hjson
Comment (vogelpi) [inline hw/ip/aes/data/aes.hjson:246]: Please add a line saying that this feature can be disabled using a compile-time Verilog parameter.

File: hw/ip/aes/data/aes.hjson
Comment (vogelpi) [reply hw/ip/aes/data/aes.hjson:304]: No, let's not drop it. But please add a line saying that this can be disabled using a compile-time Verilog parameter.

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vogelpi) [inline hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:31]: Note: two is not sufficient to fully test this. There can be three blocks in the AES engine at the same time:
- One in the data input registers waiting to get processed
- One in the engine currently being processed
- One in the data output registers waiting to get read by SW

Thus please choose a higher minimum message length, I would go for at least 4 blocks.

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vogelpi) [inline hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:40]: I appreciate that you want to give enough detail for the test implementer to properly implement the test. But here you can make your own life easier a bit by looking at the DIF and see what it provides. For example, the DIF already provides a function for this. So you can just say something like "Configure the desired operation, mode, key length and set the engine to automatic operation."

Comment (vogelpi) [reply hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:40]: Because the test implementer will use the DIF.

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vogelpi) [inline hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:44]: Again, you can just say to configure the IV. The DIF is taking care of the details you describe here.

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vogelpi) [inline hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:57]: This is not correct. You just need to write the new data to the DATA_IN registers. The output registers need to be read only to unblock the finishing of the second block (not the start). All this is described in great detail in the AES spec:  https://opentitan.org/book/hw/ip/aes/doc/programmers_guide.html#block-operation and the DIF https://github.com/lowRISC/opentitan-integrated/blob/master/sw/ip/aes/dif/dif_aes.c#L303-L355

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vogelpi) [inline hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:78]: ?

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vogelpi) [inline hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:115]: This description is clear but can be simplified a bit (see previous comments).

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vogelpi) [inline hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:42]: You also need to provide an IV for CFB and OFB modes by the way. But also here, you can make your life easier by referring to the corresponding DIF functions.

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vogelpi) [inline hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:124]: ?

**Review by vogelpi — COMMENTED** [review]
Thanks for the updates Vladimir. I've left further comments. Some of my previous questions still haven't been addressed though...

Regarding the AES.PRNG.RESEED_RATE feature: One could define a test which encrypts, say 70 blocks and after setting up (providing the key!) shuts down the entropy complex. AES should then stop / hang during block number 64 or 1 (depending on the config). Given that this test runs on the chip, we could even test 8k blocks. Here you don't need to test the AES output vs a golden model / test vectors, it's sufficient to just test that AES hangs or not.

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vrozic) [reply hw/top_earlgrey/data/ip/chip_aes_testplan.hjson]: I see. I've changed the description now

**Comment by vrozic** [comment]
The latest push contains the following changes:
    -Corrected and simplified descriptions of `chip_sw_aes_multi_block` and `chip_sw_aes_interrupt_encryption`
    -Indicated that some features can be disabled using compile-time Verilog parameter
    -Modified LC states since most tests only need to be run in PROD. One exception is `chip_sw_aes_enc` for which I've also added TEST_UNLOCKED. because it is a very basic (smoke) test. Another exception is `chip_sw_aes_sideload` which interracts with the keymgr. It is tested in `["DEV", "PROD", "PROD_END", "RMA"]` , same as the test `chip_sw_keymgr_sideload_aes` in `chip_keymgr_testplan.hjson`.
   -Modified `si_stage`. Two basic functionality tests (`chip_sw_aes_enc` and `chip_sw_aes_multi_block`) are set to SV1. Tests of simple functionality and interactions with other parts of the chip (`chip_sw_aes_sideload`, `chip_sw_aes_entropy`,  `chip_sw_aes_idle`) are set to SV2, and more complicated tests (`chip_sw_aes_interrupt_encryption`) are set to SV3. 

To do:
    - Add a test according to @vogelpi description to test `AES.PRNG.RESEED_RATE`
    - This or similar test can be used to cover  `AES.PRNG.KEY_TOUCH_FORCES_RESEED`
    - Feature `AES.PRNG.FORCE_MASKS` will probably have to remain untested

File: hw/ip/aes/data/aes.hjson
Comment (vogelpi) [reply hw/ip/aes/data/aes.hjson]: Yes, that's correct.

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vogelpi) [reply hw/top_earlgrey/data/ip/chip_aes_testplan.hjson]: This is now done, thanks.

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vogelpi) [reply hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:126]: SV2 sounds right as this tests the interaction with other blocks.

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vogelpi) [reply hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:92]: This is now better. Thanks!

**Review by vogelpi — COMMENTED** [review]
Thanks @vrozic , this looks better now. I've one or two remaining comments and you should also add the test for the PRNG reseed rate.

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vrozic) [reply hw/top_earlgrey/data/ip/chip_aes_testplan.hjson]: I've changed the test description to reflect what's been implemented

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vrozic) [reply hw/top_earlgrey/data/ip/chip_aes_testplan.hjson]: fixed

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vrozic) [reply hw/top_earlgrey/data/ip/chip_aes_testplan.hjson]: OK. I've changed it now

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vrozic) [reply hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:96]: fixed

**Comment by vrozic** [comment]
Latest few pushes contain the following changes:
-Adds description of `chip_sw_aes_prng_reseed` test that covers feature AES.PRNG.RESEED_RATE
-Adds description of `chip_sw_aes_force_prng_reseed` test that covers feature AES.PRNG.KEY_TOUCH_FORCES_RESEED
-Changes the old description of test `chip_sw_aes_enc` to reflect what's been implemented in the smoke test.
-fixes some typos
-rebase and resolves conflicts

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vogelpi) [reply hw/top_earlgrey/data/ip/chip_aes_testplan.hjson]: Cool thanks.

**Comment by vrozic** [comment]
The changes in latest push:

1. Rebase
2. Fixes typos
3. Moves `chip_sw_aes_prng_reseed` and `chip_sw_aes_force_prng_reseed` from SV2 to SV3. These tests interact with EDN, but the purpose of these tests is more complex than just checking the connection to EDN (`sw_aes_entropy` does that). Therefore, i think that SV3 is more appropriate for these tests

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vogelpi) [inline hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:147]: What should be added here is to enable the full entropy complex again (note that there is specified order regarding how to enable the EDNs, CSRNG etc. you cannot just disable EDN and then re-enable it again later) to verify that AES unstalls and finishes the last block.

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vogelpi) [inline hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:171]: Same here: re-enable the entropy complex to ensure after AES gets the entropy, it can indeed advance.

Comment (vogelpi) [inline hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:171]: Regarding verifying that the encryption stalls: one has to first provide the key and then check that the AES unit remains busy.

An encryption cannot even be started. In a mode requiring an IV (not the case here as the test tests ECB only), this IV could never be provided as AES ignores writes to the IV registers when being busy / reseeding PRNGs. This needs to be reflected in the test description.

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vogelpi) [inline hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:136]: prng -> PRNG

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vogelpi) [inline hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:163]: prng -> PRNG

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vogelpi) [inline hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:138]: It's EDN or entropy complex (comprising the EDNs, CSRNG and entropy_src), please change this here and below. Let's adhere to the same terminology used everywhere else in the project.

**Review by vogelpi — COMMENTED** [review]
Thanks for defining the additional tests @vrozic .

I have some minor corrections + the tests should be extended to verify that after the entropy complex is live again, AES can continue. Otherwise, it's not clear whether the stalls are indeed due to an entropy request or misconfiguration.

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vrozic) [reply hw/top_earlgrey/data/ip/chip_aes_testplan.hjson]: OK. I've added re-enabling of the entropy complex to the descriptions

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
Comment (vrozic) [reply hw/top_earlgrey/data/ip/chip_aes_testplan.hjson]: fixed

**Comment by vogelpi** [comment]
CHANGE AUTHORIZED: hw/ip/aes/data/aes.hjson

**Review by vogelpi — APPROVED** [review]
LGTM!

**Comment by vogelpi** [comment]
To merge this, we need another CHANGE AUTHORIZED from a committer. Would @marnovandermaas be able to do that please?

**Comment by vrozic** [comment]
Latest push only does rebase

**Comment by hcallahan-lowrisc** [comment]
CHANGE AUTHORIZED: hw/ip/aes/data/aes.hjson

**Pull request merged**
